### DISCUSSION TITLE
Spiral Simulation Running out of GPU Memory

### DISCUSSION THREAD

**Discussion by nathanjbeaumont** [JuliaHealth/KomaMRI.jl#519] [Q&A]
Hello,
I have a spiral sequence with 48 interleaves I'm trying to simulate with the brain_phantom2D. To get an image that doesn't have the screen door looking artifact I need to set `us=2` in `brain_phantom2D()`. However if I do this even with only 48 interleaves in the sequence my GPU runs out of memory. I have an RTX 2070 16gigs of GPU memory, (8 dedicated, 8 shared). Run tests.jl in the zip below to see:

[tests.zip](https://github.com/user-attachments/files/18056405/tests.zip)

![image](https://github.com/user-attachments/assets/a7376748-43c6-4e13-ac93-ecae73aef4c5)

I'm able to run it on CPU but the recon looks very bad for a fully sampled k-space:
![image](https://github.com/user-attachments/assets/4fa74b35-2ed1-4545-b210-db49c05cd1c0)

**Comment by cncastillo** [comment]
Hi!

Regarding the GPU running out of memory, you can reduce the size of the simulation blocks by increasing `sim_params["Nblocks"]`. So for example:

```julia
sim_params = KomaMRICore.default_sim_params()
sim_params["Nblocks"] = 2_000
```
Regarding the reconstruction, If I reconstruct the first 16 readouts, I get:
```julia
obj = brain_phantom2D(; us=2) 
```
<img src="https://github.com/user-attachments/assets/e6dab3c2-1537-4b48-957a-e2ee19028995" width="50%" />

I chose that number because the signal appears to have some oscilations:
![newplot2](https://github.com/user-attachments/assets/02cfc188-af93-4de5-8395-0a668402e713)

I believe this is happening because of the flip angles being too high for the acquisition type. If you are trying to create a balanced sequence, there are a few things missing:
- Add a refocusing gradient in z-direction
- Alternate flip angles (+alpha, -alpha, +alpha, ....). 

I patched the sequence by doing this (only the second point):
```julia
seq = read_seq("temp.seq")
idx = findall(is_RF_on.(seq))[2:end] # Removes IR pulse from phase alternation
for (i, idx) in enumerate(idx)
    factor = ((-1)^(i-1) + 0im)
    seq.RF[idx] = factor * seq.RF[idx]
    seq.ADC[idx+3].ϕ = angle(factor) # The phase factor in the ADC is important
end 
```
and I get something that looks much better
<img src="https://github.com/user-attachments/assets/fefd43ec-a8a9-4b75-b010-00e0e6a81178" width="50%"/>

As a side note I would avoid to use a slice-selective RF pulse for the initial inversion, it is much better to use an adiabatic hyperbolic secant inversion. Let me know if you need help implementing it, but there are some examples in the `examples` folder.

**Reply by nathanjbeaumont** [reply]
Thanks for the response.

> Add a refocusing gradient in z-direction

Do you mean a rewinder gradient in z after the RF pulse? There is already a rewinder after every RF pulse:
![image](https://github.com/user-attachments/assets/3057df0e-ed6e-4b85-8e83-c53530f2e378)

> Alternate flip angles (+alpha, -alpha, +alpha, ....).

The [paper I'm trying to reproduce](https://pmc.ncbi.nlm.nih.gov/articles/PMC4461545/) uses what they are calling "FISP". and they say:

> The long-used Fast Imaging with Steady State Precession (FISP) sequence acquires coherent steady state signals with a constant unbalanced gradient moment in each repetition time

So I was assuming they used the a phase of 0 for every RF pulse. However [this article](https://mriquestions.com/grassfisp-how.html#/) claims FISP uses alternating pulse phases. [This article also says](https://mriquestions.com/grass-details.html#/):

> rapidly applied RF-pulses result in saturation of the longitudinal magnetization. If the RF-pulses were continuously applied in the same (+α) direction, this steady-state magnetization is lower

**Reply by cncastillo** [reply]
Oh sorry, I thought you wanted to recreate a trueFISP, aka balanced Steady State Free Precession (bSSFP). https://pmc.ncbi.nlm.nih.gov/articles/PMC3602925/

The additional gradient in z can be seen here:
![image](https://github.com/user-attachments/assets/b62e6d67-71b6-4aab-9be6-7a5c9f15f50c)

In that case, with MRF-FISP you generally use spoiled gradient echo (https://www.sciencedirect.com/science/article/pii/S1936878X18308337):
>  The cMRF pulse sequences used to date have a fast imaging with free precession (FISP) readout with no radiofrequency phase cycling, instead of the originally proposed bSSFP readout. Unlike the bSSFP readout, FISP uses an unbalanced spoiler gradient in the slice select direction, thus making this sequence relatively insensitive to off-resonance effects.

For the simulation, to have a realistic gradient spoiling, you need to add spins in the z dimension or use `brain_phantom3D` (you can use the `start_end` keyword argument to adjust the thickness and `us` for the up-sampling).

**Reply by nathanjbeaumont** [reply]
[The paper also says this](https://pmc.ncbi.nlm.nih.gov/articles/PMC4461545/):

> The FISP sequence does not lead to the banding artifacts that are seen in bSSFP. Without any other mechanisms to destroy the coherence of the transverse magnetization (such as RF spoiling), the sequence is sensitive to both longitudinal and transverse relaxation times.

So it sounded to me like they didn't use spoilers. They're purposefully trying to keep transverse magnetization. Also you can see they rewind back to the center of k-space after the spiral, which would be pointless if they're going to spoil right after the rewind right?

**Reply by aTrotier** [reply]
Take a look at this youtube video : https://youtu.be/6wAuxKf1GhQ?si=UQ06l2gw9kKJe_Ib

The "spoiler" needs to be constant for each TR (and no RF spoiling) in order to obtain a FISP sequence

**Reply by nathanjbeaumont** [reply]
OK I believe I figured out the issue, the paper says:

> Figure 1a shows one example of the MRF-FISP sequence with the unbalanced gradient along the slice-selection direction. In this study, we used a SINC waveform RF pulse with 2 ms duration. The time bandwidth product of the RF pulse was set to 8 to minimize the imperfect slice profile. The unbalanced gradient moment in the slice-selection direction achieved 8π dephasing across 5 mm slice thickness within each repetition time.

So I believe the rephasing gradient after slice selection had an area _less than_ half the slice selection gradient so that the rephasing was incomplete (8π of dephasing remaining).

**Comment by aTrotier** [comment]
No they said that their is a net moment of 8pi at the end of the sequence
along the slice direction.

You must rephase everything along read and phase and just put an
additional spoiler at the end of the TR along the slice dimension of 8pi
area.

 If you put the de phasing gradient of 8pi after the slice selection you will obtain something like a PSIF sequence I think.

**Reply by cncastillo** [reply]
This! :arrow_up: 

An $8\pi$ spoiler along the slice direction is what would make sense to me. The phase dispersion of this spoiler is dependent on the slice thickness $\Delta z$, so you need to calculate the area of the gradient $A_{\mathrm{grad}}$ such that 

$$
\phi = \gamma A_{\mathrm{grad}} \Delta z = 8\pi.
$$

**Comment by nathanjbeaumont** [comment]
@cncastillo I would like to use an adiabatic pulse for the inversion at the beginning as that's what the paper does, but I don't know how to design such a pulse. I see `examples\3.tutorials\adiabatic_pulses\BIR4.mat`. How was BIR4.mat generated? Is the script for that available?

**Reply by cncastillo** [reply]
I think I lost the script that generates it unfortunately, but at the top of the scripts I put the name of the paper used as reference. I believe that in the paper you are trying to replicate they use a hyperbolic secant (HS) pulse. These are some examples:
- BIR4: https://github.com/JuliaHealth/KomaMRI.jl/blob/master/examples/3.tutorials/adiabatic_pulses/BIR4InversionProfile.jl
- HS: https://github.com/JuliaHealth/KomaMRI.jl/blob/6d4deb8a1d6548da25cc6973080a6d94fc17be0b/examples/4.reproducible_notebooks/pluto-03-low-field-boost-optimization.jl#L188-L215
- HS: https://github.com/JuliaHealth/KomaMRI.jl/blob/master/examples/3.tutorials/adiabatic_pulses/HSInversionProfile.jl

To be honest some of this are almost ready to be a Literate example. @Stockless could you create one based on HSInversionProfile.jl?

**Reply by cncastillo** [reply]
I found some Mathematica code to generate it that might be useful:
![image](https://github.com/user-attachments/assets/b8b22845-70b9-42f4-b554-eded33e9a292)
You may want to adjust the $\beta$ and $\kappa$ parameters.

**Reply by aTrotier** [reply]
I am using the sigpy python package for that (already implemented in pypulseq)

**Reply by aTrotier** [reply]
Take a look at : https://sigpy.readthedocs.io/en/latest/_modules/sigpy/mri/rf/adiabatic.html#bir4
